### ISSUE TITLE
TF batch gradient calculation

### ISSUE THREAD

**Issue by llodds** [tensorflow/tensorflow#49865]
I am trying to figure out the math model (or data flow) to compute gradient for batch gradient descent algorithms. All the materials I have been reading seems to say that the final gradient of a batch is the average of all gradients evaluated for all training samples of a batch, but when I check the TF loss function of a batch, I found the output of a TF loss function is a scalar value rather than (BATCH_SIZE,) -- a scalar vector of length equals to BATCH_SIZE. So I am wondering how TF calculates the gradient of a batch, does it just use this scalar loss value to calculate the gradient only ONCE for the entire batch on the trainable variables, or this loss value is used by BATCH_SIZE times to calculate the gradients for each training pairs of a batch, and then it does an average. 

Any reference is welcome. Thanks.

**Comment by tilakrayal** [comment]
@llodds ,

Please refer these link for the information on gradient.[link1](https://www.oreilly.com/library/view/getting-started-with/9781786468574/ch02s03.html),[link2](https://www.tensorflow.org/quantum/tutorials/gradients),[link3](https://www.tensorflow.org/guide/advanced_autodiff).

Please file an issue in [Stack Overflow](https://stackoverflow.com/questions/tagged/tensorflow), as it is neither a bug nor a feature request.

Thanks!

**Comment by llodds** [comment]
Thanks!
